### PR TITLE
Completes the resolution of document module supertypes

### PR DESCRIPTION
**Pull request by retailcoder**
This PR completes the work done on supertype resolution.

An unqualified `Cells` member call resolves to `Excel._Global.Cells` same as before in, say, `ThisWorkbook`:

_Global.Cells

But now in `Sheet1` the same identical code is now resolving to `Excel._Worksheet.Cells`:

_Worksheet.Cells

The *implicit ActiveSheet* and *implicit containing sheet* reference inspections still work as intended after including the `_Worksheet` superclass in the class names array.

Interestingly it's the `Worksheet` type that comes up on qualified instance member calls:

_Worksheet members of a document module resolve to Worksheet members when qualified

As a result, `Sheet1.Cells` resolves to a member of `Excel.Worksheet`, as it should:

Sheet1.Cells property

### CODE REVIEW

**Comment by MDoerner** [comment]
Could you explain why you add a global variable for `_Worksheet`? There should already be one from the reference to `Excel` and that should be taken up by the super type resolution. Moreover, why do you want to add variable declarations for document modules? 

To me the original issue of the `Cells` calls not resolving correctly looks like an issue in the name resolution regarding supertypes if there is also a global object with the same member. Have you looked into that already and determined that there is no way to get this working cleanly with the already existing `_Worksheet`?

File: Rubberduck.CodeAnalysis/Inspections/Abstract/ImplicitSheetReferenceInspectionBase.cs [hunk 1]
```diff
@@ -35,7 +36,8 @@ protected override IEnumerable<Declaration> ObjectionableDeclarations(Declaratio
 
         private static readonly string[] GlobalObjectClassNames =
         {
-            "Global", "_Global"
+            "Global", "_Global", 
+            "Worksheet", "_Worksheet"
```
Comment (bclothier) [inline Rubberduck.CodeAnalysis/Inspections/Abstract/ImplicitSheetReferenceInspectionBase.cs:40]: This feels wrong. 1) we should be able to enumerate document modules already, and infer the base class without needing to check a list of constants. As it is, it won’t port to other hosts. 2) by definition, the appobjects are predeclared, but appobjects are different in other hosts. We can then infer Excel’s Globals because it’s an appobject.

To avoid needing resolving the interface to a coclass using strings we should enumerate coclasses’ implemented interfaces and thus locate the correct coclass.

Comment (retailcoder) [reply Rubberduck.CodeAnalysis/Inspections/Abstract/ImplicitSheetReferenceInspectionBase.cs:40]: This inspection is very much Excel-specific though. Agreed the inspection feels like it's upside-down and needs adjustment.

Comment (bclothier) [reply Rubberduck.CodeAnalysis/Inspections/Abstract/ImplicitSheetReferenceInspectionBase.cs:40]: Looking at this again, I see the code is actually in inspection - I was conflating that with the `Parsing` project. Still, I think it should be the `Parsing` project that provides the necessary metadata for the inspection to then work upon. the inspection can/should filter for specific Excel coclasses/interfaces, but it should be doing off a list of some kind, not a bunch of constant strings.

Comment (retailcoder) [reply Rubberduck.CodeAnalysis/Inspections/Abstract/ImplicitSheetReferenceInspectionBase.cs:40]: Absolutely: with properly-resolving unqualified `Workbook` and `Worksheet` members, it should be possible to identify implicit `ActiveWorkbook` and implicit `ActiveSheet` references by simply pulling the identifier references of `_Global`.

**Comment by retailcoder** [comment]
> Could you explain why you add a global variable for _Worksheet? There should already be one from the reference to Excel and that should be taken up by the super type resolution.

Adding `_Worksheet` to the inspection's `GlobalObjectClassNames` array is wrong, but it was the simplest way to make the inspection fetch the declaration for the `_Worksheet` interface for which the member calls are resolving now that supertypes correctly get added to the document modules ([this diff](https://github.com/rubberduck-vba/Rubberduck/pull/5721/files#diff-03e847f452261201360012d521b3f4db06e43245de9f0ab7fe1714e3af09d71fL937-R938) is what's fixing that, that condition wasn't being entered): `_Worksheet` and `_Workbook` needed to be added because the string array is defined in a base class from which several inspections are derived; it was needed to keep the "implicit containing work{sheet|book}" inspections working, because _containing workbook/worksheet_ refs that should have never been resolving to `_Global.MemberName` in the first place are now correctly resolving to `_Worksheet.MemberName` as per the document module's type hierarchy.

With the supertypes correctly applying, the "implicit activesheet/activeworkbook ref" inspections can probably now just flag all references to `_Global.{WorksheetMemberName}`, and assuming qualified calls resolve to `Worksheet` and the unqualified calls resolve to `_Worksheet` (it appears to be the case and it's rather convenient!), then the "containing workbook/worksheet ref" inspections can now just flag all references to `_Worksheet.{WorksheetMemberName}`.

> Moreover, why do you want to add variable declarations for document modules? To me the original issue of the Cells calls not resolving correctly looks like an issue in the name resolution regarding supertypes if there is also a global object with the same member.

Because VBA _supposedly_ defines global variables named after the document modules, so in principle there's a `ThisWorkbook` and a `Sheet1` identifier in project scope that's legal to use in VBA code, but we are not currently resolving it - it's the elusive "code name" identifier.

Without that declaration, qualified `Sheet1` and `ThisWorkbook` calls to `Worksheet` or `Workbook` members are failing resolution altogether. I agree it's very much not ideal.

If document modules had a PredeclaredId, we would already be picking it up - *and we don't*, so I'm guessing the `Sheet1` dynamic compile-time global variable must exist in the project's `ITypeLib` and that *would* indeed be the place to get it *if it's there*; I did not try that approach, instead I admittedly cheated and did essentially what we do with controls on `UserForm` designers: we create a "fake" declaration for a public field, and the resolver does the rest. Here we create a "fake" declaration named after the document module, and the resolver does the rest. If/when that global identifier is created off the `ITypeLib` then we don't need the "fake" declaration anymore.

Alternatively we could have document modules have a fake PredeclaredId "attribute"; I think that might actually be cleaner than declaring a fake project-scope variable - we'd resolve `Sheet1` to `VBAProject.Sheet1 (class)` instead of `VBAProject.Sheet1.Sheet1 (variable)`, ...which would conveniently avoid breaking the *shadowed declaration* inspection/tests, wouldn't introduce an extra declaration, and would remove the eventual need to get it from the `ITypeLib` and start breaking the *shadowed declaration* inspection in a way that doesn't strike me as one that would necessarily break the tests for it.

Bottom line, in order to correctly resolve `Sheet1` member calls we're missing a declaration *somewhere*, whether it's a PredeclaredID, a fake project-scope variable, or something else.

I'm going to be trying the fake PredeclaredID approach, I think it's more sensible than trying to get a variable from the `ITypeLib`, because the user-facing representation of that approach feels much more intuitive than some magical `VBAProject.Sheet1.Sheet1` variable - whether the magical variable actually exists or not.

**Comment by MDoerner** [comment]
Regarding the variable, I think we should simply set the `PredeclaredId` attribute on document modules. Actually, we should take a look what is going on with the attributes pass. I have the feeling that we use the version of the file export for the export command in the parsing as well. Looking at the result of an exported worksheet, I see `PredeclaredId = True` and `Exposed = True`.

**Comment by retailcoder** [comment]
Oooooh my... hold on, so the only reason we're not picking up a `VB_PredeclaredId` on document modules is because we're not importing them back for the attributes pass, right? If that's so, then yeah - I'm ditching the fake variable and adding a fake `VB_PredeclaredId` attribute for document modules. Yay!!

**Comment by retailcoder** [comment]
@MDoerner `SimpleNameDefaultBinding` was classifying the default instance variable as a `Type`; seems changing it to resolve as a `Variable` fixes the problem - I have removed the clumsy fake variable declaration.

**Review by bclothier — APPROVED** [review]

File: Rubberduck.Parsing/VBA/DeclarationCaching/DeclarationFinder.cs [hunk 2]
```diff
@@ -357,6 +357,17 @@ public ICollection<Declaration> FindFormEventHandlers()
...
         public IEnumerable<Declaration> Projects => _projects.Value;
 
+        /// <summary>
+        /// Gets the <see cref="ProjectDeclaration"/> object for specified referenced project/library.
+        /// </summary>
+        /// <param name="name">The identifier name of the project declaration to find.</param>
+        /// <param name="result">The <see cref="ProjectDeclaration"/> result, if found; null otherwise.</param>
+        public bool TryFindProjectDeclaration(string name, out Declaration result)
+        {
+            result = _projects.Value.SingleOrDefault(project => project.IdentifierName.Equals(name, StringComparison.InvariantCultureIgnoreCase) && !project.IsUserDefined);
```
Comment (MDoerner) [inline Rubberduck.Parsing/VBA/DeclarationCaching/DeclarationFinder.cs:367]: Can we make this `FirstOrDefault` just in case? As it is, this might throw, if somebody names a referenced locked project with a name clash.

Comment (retailcoder) [reply Rubberduck.Parsing/VBA/DeclarationCaching/DeclarationFinder.cs:367]: Ah, good call - I'll also parameterize the `IsUserDefined` check, makes it more versatile. Currently this method has only the two callers it was extracted from.
